### PR TITLE
Validate when reorg ends

### PR DESCRIPTION
**Pull request by limebell**
Validate function must be passed to `ValidatingActionRenderer`.

### CODE REVIEW

**Comment by limebell** [comment]
Validate function became not to be passed to its constructor, it's now built in. Instead, `BlockChain` property should be set before any reorg rendering.

**Review by earlbread — APPROVED** [review]

File: Lib9c/Renderer/ValidatingActionRenderer.cs [hunk 1]
```diff
@@ -92,51 +89,105 @@ public override void RenderBlockEnd(Block<T> oldTip, Block<T> newTip)
...
-                expectedActions.AddRange(tx.Actions.Cast<IAction>());
+                repeat = !block.PreviousHash.Equals(branchpoint.Hash);
+                expectedUnrenderedActions.AddRange(
+                    block.Transactions.SelectMany(t => t.Actions).Cast<IAction>());
+                if (!(chain.Policy.BlockAction is null))
+                {
+                    expectedUnrenderedActions.Add(chain.Policy.BlockAction);
+                }
+
+                block = block.PreviousHash is null ? chain.Genesis : store.GetBlock<T>(block.PreviousHash.Value);
```
Comment (longfin) [inline Lib9c/Renderer/ValidatingActionRenderer.cs:130]: I guess `block` is already `chain.Genesis` if `block.PreviousHash is null`. 😕

File: Lib9c/Renderer/ValidatingActionRenderer.cs [hunk 2]
```diff
@@ -92,51 +89,105 @@ public override void RenderBlockEnd(Block<T> oldTip, Block<T> newTip)
...
-                expectedActions.Add(_policy.BlockAction);
+                repeat = !block.PreviousHash.Equals(branchpoint.Hash);
+                var actions = block.Transactions.SelectMany(t => t.Actions).Cast<IAction>().ToList();
+                if (!(chain.Policy.BlockAction is null))
+                {
+                    actions.Add(chain.Policy.BlockAction);
+                }
+
+                expectedRenderedActions = actions.Concat(expectedRenderedActions).ToList();
+                block = block.PreviousHash is null ? chain.Genesis : store.GetBlock<T>(block.PreviousHash.Value);
```
Comment (longfin) [inline Lib9c/Renderer/ValidatingActionRenderer.cs:145]: I guess `block` is already `chain.Genesis` if `block.PreviousHash is null`. 😕

File: Lib9c/Renderer/ValidatingActionRenderer.cs (hunk 1 above)
Comment (limebell) [reply Lib9c/Renderer/ValidatingActionRenderer.cs]: You are right. Amended in [`9e44d67`](https://github.com/planetarium/lib9c/pull/124/commits/9e44d67e40b808fda503a4f58e86a3bf96dae280).

**Review by longfin — APPROVED** [review]

**Pull request merged**